clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -972,16 +975,28 @@ private Record nextFetchedRecord() {
         private List<ConsumerRecord<K, V>> fetchRecords(int maxRecords) {
             if (isFetched)
                 return Collections.emptyList();
+            if (nextInlineException != null) {
```

Comment:
The point is not that the next record is at a higher offset, but that the current position deterministically determines the next record to be returned. After we have caught the exception from the record at offset 1 (say), the position remains at offset 1. However, the next fetched record may be at offset 2. If we catch an error at that offset as well, the position will remain at 1 (I guess?) and the next record we try to return may be at offset 3. This makes the consumer less functional since the behavior of a fetch at the current offset depends on the history of previous fetches. 

The alternative is that the consumer always fetches from the current position. If the user wants to skip a message, they must explicitly seek to the next larger offset. Then the result from a poll is always deterministic. The user is in control and the consumer will not try to guess what the user wants to do.

Whether by design or by accident, this is how the new consumer has worked up until now, so another thing to consider is whether changing this behavior will cause any compatibility issues.